Add support for ElasticSearch alias to read spans from (Resolves #68)

frittentheke opened the pull request:
## Which problem is this PR solving?
Currently the spark job expects the spans to always come from date-postfixed indices, even though the  Jaeger Collector can be configured to use ElasticSearch aliases and to make use of the rollover API (see:https://github.com/jaegertracing/jaeger/pull/1309)

If that is the case the dependencies can simply not be extracted.
See issue: https://github.com/jaegertracing/spark-dependencies/issues/68

## Short description of the changes
* Added a ES_USE_ALIASES parameter - just like the other Jaeger components have
* Added a range query to only return the spans of the last 24 hours (lust like per day index would do)



Creds also go to my mate @msp14 who patiently explained Spark to me so I was able to create this PR.

pavolloffay commented:
@frittentheke could you please sign off the commits?

frittentheke commented:
@pavolloffay  sorry about that, missed this on the second commit I pushed this morning.
Should all be signed-off now.

pavolloffay commented on jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java:
```diff
@@ -186,6 +188,8 @@ private static String getSystemPropertyAsFileResource(String key) {
       conf.set(entry.getKey(), entry.getValue());
     }
     this.indexPrefix = builder.indexPrefix;
+
```
nit: remove empty line

frittentheke replied:
Gladly removed nothing ;-)

pavolloffay commented on jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java:
```diff
@@ -219,7 +228,13 @@ void run(String[] spanIndices, String[] depIndices,String peerServiceTag) {
+        // When using rollover indices there might be none or even multiple per day - thus Jaeger expects an alias to be available
+        // pointing to all potential indices.
+        // Always applying this "filter" does not do any harm even for the indexDated ones
+        String esQuery = "{\"range\": {\"startTimeMillis\": { \"gte\": \"now-24h\" }}}";
```
+1 this looks good.

frittentheke commented:
@pavolloffay  FTAL (again).

christine-gong commented:
@frittentheke I'm waiting on this feature! thanks for your effort! I saw your tests did pass on es7. is there any diff in es7 and non-es7 that missed here?

frittentheke commented:
@christine-gong I just tried to replicate the issues locally, but apparently the error is not solid. I had success sometimes. Maybe some sort of timing issues / race condition somewhere?

I updated the ES version from 5.x to 6.8.8 as part of commit  https://github.com/jaegertracing/spark-dependencies/pull/86/commits/48b6321d37526b53d722f24ce33565c1851499b0 though, as ES 5.x is End of Life for a while now (https://www.elastic.co/pt/support/eol)

pavolloffay commented:
Let's keep ES5 compatibility and be aligned with Jaeger ES version support. I don't think there is anything that should prevent us from using ES5.

frittentheke commented:
> Let's keep ES5 compatibility and be aligned with Jaeger ES version support. I don't think there is anything that should prevent us from using ES5.

@pavolloffay  allright - I changed that commit to a simple update to the latest point releases for ES 5 and 7.

frittentheke commented:
@pavolloffay do you mind having a look at the failing CI jobs.  Looking at my changes again - they don't seem to affect the code path for the tests.
I also replicated them locally and they succeed or fail randomly there as well. This seems to be the case for previous jobs (not related to this PR) as well (https://travis-ci.org/github/jaegertracing/spark-dependencies/jobs/572258243).

christine-gong commented:
@pavolloffay @frittentheke another thing I observed when I tried this changeset locally: jaeger ui uses api/dependencies to fetch dependencies while https://github.com/jaegertracing/jaeger/blob/5ed586c114152dfdb9f9643b9303fd40d45cc6e5/plugin/storage/es/dependencystore/storage.go#L86 relies on dated dependency indices too. To make it work with the minimum change, i updated the `(prefix)-jaeger-dependencies-write` to dated one `(prefix)-jaeger-dependencies-yyyy-mm-dd`

frittentheke commented:
@christine-gong I also created a two PRs for the dependency store ... 

* https://github.com/jaegertracing/jaeger/pull/2144 
* https://github.com/jaegertracing/jaeger/pull/2149

christine-gong commented:
@frittentheke got it! Sorry I missed that! Will check that tomorrow! Btw would current solution be able to run multiple times per day? E.g. I ran the spark job twice in a hour, Ui showed the summed up call counts instead of the latest one (time stamp fields in es both were today’s date)

Leinad0033 commented:
Any chance of this getting updated to a state where is can be merged into master?

frittentheke commented:
> Any chance of this getting updated to a state where is can be merged into master?

@pavolloffay ?

alexanderccc commented:
Are there any updates on this? I've also started using ILM with our jaeger indices in elasticsearch which made the dependencies job unusable as they're searching for data in the timestamped indices. Are there any plans on finishing this and having It integrated in the dependency jobs?

frittentheke commented:
> The prerequisite to merge this is to implement aliases support for the dependency index in the main repo [jaegertracing/jaeger#2144](https://github.com/jaegertracing/jaeger/pull/2144).
> 

> As in intermediate solution the dependencies could be stored into daily indices to keep the current compatibility and then once [jaegertracing/jaeger#2144](https://github.com/jaegertracing/jaeger/pull/2144) lands add a second conf property to store the results to the write alias. If we go this route we should rename `ES_USE_ALIASES` to include span name e.g. `ES_USE_SPAN_READ_ALIAS` or just `ES_USE_READ_ALIAS`.
> 
> The PR has to be rebased.

I added `ES_USE_READ_ALIAS` and `ES_USE_WRITE_ALIAS` with two individual commits now.
This allows for the intermediate solution of reading from an alias and independently switch to using a write alias.
I also rebased to current master.

Could you kindly PTAL @pavolloffay .


@alexanderccc FYI ^^

Leinad0033 commented:
@pavolloffay Any possibility of getting a review for this?

Korzi commented:
Hi, is it going to be reviewed/merged? Or is there any other solution for spark job to use aliases?

frittentheke commented:
@Leinad0033 @Korzi where you able to verify if my changes work for you then?

@pavolloffay I know you are the only one maintaining this repo, but I just refreshed my PR https://github.com/jaegertracing/jaeger/pull/2144 and maybe we can get them both reviewed one merged. They have been open so long - and both only add the little bit of functionality to be able to use ES aliases.

frittentheke commented:
> @pavolloffay I know you are the only one maintaining this repo, but I just refreshed my PR [jaegertracing/jaeger#2144](https://github.com/jaegertracing/jaeger/pull/2144) and maybe we can get them both reviewed one merged. They have been open so long - and both only add the little bit of functionality to be able to use ES aliases.

@pavolloffay  @albertteoh with https://github.com/jaegertracing/jaeger/pull/2144 now merged, could you maybe PTAL at this PR here again? With all ES indices behind rollover aliases it just makes sense to support this for the reporting as well.

albertteoh commented on jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java:
```diff
@@ -209,7 +215,27 @@ private static String prefix(String prefix) {
+    String[] writeIndex;
+
+    // read spans from alias index name
+    if (this.useReadAlias) {
```
The indentation looks inconsistent from here down.

frittentheke replied:
fixed.

albertteoh commented on jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java:
```diff
@@ -209,7 +215,27 @@ private static String prefix(String prefix) {
-    run(indexDate("jaeger-span"), indexDate("jaeger-dependencies") ,peerServiceTag);
+
+    String[] readIndices;
+    String[] writeIndex;
```
nit: `writeIndices` since it's an array?

frittentheke replied:
True but also not really intuitive then.

The data type is an array, just because the `indexDate ` helper does return an array (https://github.com/jaegertracing/spark-dependencies/blob/a41f1734c1ad647f3b501194fd49b12161780c41/jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java#L215). But this is only for read indices.

I gladly clean this up by either splitting the helpers for read and write or by removing (breaking) compatibility with the old `prefixBefore19` format, but it's changing more than the feature I wanted to add. Just let me know that you'd rather like

albertteoh commented:
@frittentheke looks like there are merge conflicts that need addressing.

frittentheke commented:
@albertteoh  ... Thanks for your lightning fast response and remarks!

To be brutally honest: I did not bother to take another look at the code (yet), until I knew you kindly would look at it.
I shall be fixing those issues and rebase the PR later.

albertteoh reviewed (approved):
LGTM.

AhHa45 commented:
bump @pavolloffay

Karth54 commented:
Eagerly awaiting the release of this feature!

phoebe-canva commented on jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java:
```diff
@@ -209,7 +215,27 @@ private static String prefix(String prefix) {
+
+    // write dependencies to alias index name
+    if (this.useWriteAlias) {
+      writeIndex = new String[] {prefix(indexPrefix) + "jaeger-dependencies-write"};
```
Heya, just raising a bug I found from this - I enabled readAlias and writeAlias and realised that when it finds no dependency on the first loop, it gives me this error
```
 22/04/13 20:05:46 INFO ElasticsearchDependenciesJob: Running Dependencies job for 2022-04-13T00:00Z, reading from jaeger-span-read index, result storing to jaeger-dependencies-write     │
│ 22/04/13 20:05:56 INFO ElasticsearchDependenciesJob: Done, 0 dependency objects created                                                                                                   │
│ Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
```

It is because `writeIndex` is one element in the array but `readIndices` has two.

Will need to do 
```suggestion
      writeIndex = new String[] {prefix(indexPrefix) + "jaeger-dependencies-write", prefixBefore19(indexPrefix) + "jaeger-dependencies-write"};
```

frittentheke replied:
Thanks for your feedback and the suggested change.
I, as you might have guessed, did not use the older prefixBefore19 style so I did not run into this issue.
But fair point. I rebased my PR and added your suggested change.

At some point it might be sensible to drop support for two styles and cleanup the code in this regard.
@albertteoh Do you mind taking another look and maybe provide a path to have this whole PR merged?

frittentheke commented:
@pavolloffay @albertteoh could you kindly take another peek at this and either suggest more changes, merge it or close it?

Karth54 commented:
In an attempt to get this working for our production jaeger instance, I took your fork @frittentheke, and created and published a separate [public docker image](https://hub.docker.com/r/karth54/jaeger-es-alias-spark-dependencies/tags) to use as a temporary solution. I didn't pass anything special when I built it. After deploying via helm, the job has been running for over 15 hours now. The environment variables that are passed are:
```
STORAGE=elasticsearch
ES_NODES=http://elasticsearch-master:9200
ES_NODES_WAN_ONLY='false'
ES_USE_READ_ALIAS='true'
ES_USE_WRITE_ALIAS='true'
```

Am I missing something? The only logs I see in the cronjob are:
```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.apache.spark.unsafe.Platform (file:/app/jaeger-spark-dependencies-0.0.1-SNAPSHOT.jar) to method java.nio.Bits.unaligned()
WARNING: Please consider reporting this to the maintainers of org.apache.spark.unsafe.Platform
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
22/04/30 22:50:10 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
22/04/30 22:50:10 INFO ElasticsearchDependenciesJob: Running Dependencies job for 2022-04-30T00:00Z, reading from jaeger-span-read index, result storing to jaeger-dependencies-write
```

frittentheke commented:
@albertteoh @pavolloffay  ... any chance somebody can get this PR merged?
Is more and more people using this by manually applying this really the right way to go?

pavolloffay commented on jaeger-spark-dependencies-elasticsearch/src/main/java/io/jaegertracing/spark/dependencies/elastic/ElasticsearchDependenciesJob.java:
```diff
@@ -60,6 +60,8 @@ public static final class Builder {
     String indexPrefix = Utils.getEnv("ES_INDEX_PREFIX", null);
     String spanRange = Utils.getEnv("ES_TIME_RANGE", "24h");
+    Boolean useReadAlias = Boolean.parseBoolean(Utils.getEnv("ES_USE_READ_ALIAS", "false"));
+    Boolean useWriteAlias = Boolean.parseBoolean(Utils.getEnv("ES_USE_WRITE_ALIAS", "false"));
```
Do we need two env variables for read and write? I would say that if read alias is used then write should be used as well. 

Also please document this in the root readme

frittentheke replied:
Done. PTAL.

frittentheke commented:
@pavolloffay  PTAL

frittentheke commented:
@albertteoh @pavolloffay  is there anybody else who you take a peek. I'd really love to have this off my list.
